File: benchmarking/benchmarks/SqueezeNet/implementations/python/squeezenet.py
from keras.layers import Input, Convolution2D, MaxPooling2D, Activation, concatenate, Dropout, warnings
from keras.layers import GlobalAveragePooling2D
from keras import optimizers
from keras.models import Model
from keras.datasets import cifar10
from keras.preprocessing.text import one_hot
import tensorflow as tf
import numpy as np

"""""""""
CONSTANTS
"""""""""
TRAIN_SIZE = 50000
TEST_SIZE = 10000
BATCH_SIZE = 64
TRAIN_STEPS = 1

sq1x1 = "squeeze1x1"
exp1x1 = "expand1x1"
exp3x3 = "expand3x3"
relu = "relu_"


class SqueezeNet:

    def __init__(self, backend):
        self.train_images = []
        self.train_labels = np.zeros((TRAIN_SIZE, 10))
        self.test_images = []
        self.test_labels = np.zeros((TEST_SIZE, 10))

        self.load_data()
        self.backend = backend

        if backend == 'cpu':
            self.device = '/cpu:0'
        else:
            assert backend == 'gpu', 'Invalid backend: %s' % backend
            self.device = "/device:GPU:0"

        self.model = SqueezeNet.create_model()

    # Modular function for Fire Node
    @staticmethod
    def fire_module(x, fire_id, squeeze=16, expand=64):
        s_id = 'fire' + str(fire_id) + '/'

        channel_axis = 3

        x = Convolution2D(squeeze, (1, 1), padding='valid', name=s_id + sq1x1)(x)
        x = Activation('relu', name=s_id + relu + sq1x1)(x)

        left = Convolution2D(expand, (1, 1), padding='valid', name=s_id + exp1x1)(x)
        left = Activation('relu', name=s_id + relu + exp1x1)(left)

        right = Convolution2D(expand, (3, 3), padding='same', name=s_id + exp3x3)(x)
        right = Activation('relu', name=s_id + relu + exp3x3)(right)

        x = concatenate([left, right], axis=channel_axis, name=s_id + 'concat')
        return x

    # Original SqueezeNet from paper.
    @staticmethod
    def create_model(input_shape=(32, 32, 3)):

        img_input = Input(shape=input_shape)

        x = Convolution2D(64, (2, 2), strides=(1, 1), padding='same', name='conv1')(img_input)
        x = Activation('relu', name='relu_conv1')(x)
        x = MaxPooling2D(pool_size=(2, 2), strides=(2, 2), name='pool1')(x)

        x = SqueezeNet.fire_module(x, fire_id=2, squeeze=16, expand=64)
        x = SqueezeNet.fire_module(x, fire_id=3, squeeze=16, expand=64)
        x = MaxPooling2D(pool_size=(2, 2), strides=(2, 2), name='pool3')(x)

        x = SqueezeNet.fire_module(x, fire_id=4, squeeze=32, expand=128)
        x = SqueezeNet.fire_module(x, fire_id=5, squeeze=32, expand=128)
        x = MaxPooling2D(pool_size=(2, 2), strides=(2, 2), name='pool5')(x)

        x = SqueezeNet.fire_module(x, fire_id=6, squeeze=48, expand=192)
        x = SqueezeNet.fire_module(x, fire_id=7, squeeze=48, expand=192)
        x = SqueezeNet.fire_module(x, fire_id=8, squeeze=64, expand=256)
        x = SqueezeNet.fire_module(x, fire_id=9, squeeze=64, expand=256)
        x = Convolution2D(10, (1, 1), strides=(1, 1), padding='same', name='conv2')(x)

        x = GlobalAveragePooling2D()(x)

        inputs = img_input

        model = Model(inputs, x, name='squeezenet')

        optimizer = optimizers.adam(lr=0.001)
        model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
        return model

    def load_data(self):
        (self.train_images, train_labels), (self.test_images, test_labels) = cifar10.load_data()

        train_labels = train_labels.flatten()
        test_labels = test_labels.flatten()

        self.train_labels[np.arange(TRAIN_SIZE), train_labels] = 1
        self.test_labels[np.arange(TEST_SIZE), test_labels] = 1

    def train(self, train_steps=TRAIN_STEPS):
        with tf.device(self.device):
            self.model.fit(self.train_images, self.train_labels, batch_size=BATCH_SIZE, epochs=train_steps)

    def predict(self):
        with tf.device(self.device):
            self.model.evaluate(x=self.test_images, y=self.test_labels, batch_size=64)


def init(backend):
    return SqueezeNet(backend)

File: benchmarking/benchmarks/SqueezeNet/javascript/squeezenet.js
import * as tf from '@tensorflow/tfjs';
import {CIFAR10} from "./cifar-10/cifar10-client";

/**
 * CONSTANTS
 */
// Hyper-parameters
const LEARNING_RATE = .001;
const BATCH_SIZE = 64;
const TRAIN_STEPS = 1;

// Data constants
const TRAINING_SIZE = 1001;
const TEST_SIZE = 1001;
const IMAGE_SIZE = 32;
const IMAGE_DEPTH = 3;

let data = {};

/**
 * Loads data into memory
 * @return {Promise<void>}
 */
async function loadData(training_size, test_size) {
    const cifar = new CIFAR10();

    const training = await cifar.training.get(training_size);

    const test = await cifar.test.get(test_size);

    data = {
        train: {
            images: training.map((obj) => obj.input),
            labels: training.map((obj) => obj.output),
            num_images: training.length,
        },
        test: {
            images: test.map((obj) => obj.input),
            labels: test.map((obj) => obj.output),
            num_images: test.length,
        }
    };
}


/**
 * SqueezeNet Model
 */
export class SqueezeNet {

    constructor(){
        this.model = SqueezeNet.create_model();
    }

    /**
     * Creates and compilers tensorflow model
     * @return {Model}
     */
    static create_model(){
        const optimizer = tf.train.adam(LEARNING_RATE);
        const input = tf.input({shape: [IMAGE_SIZE, IMAGE_SIZE, IMAGE_DEPTH]});
        const conv1 = SqueezeNet.create_conv(96, 2, input);
        const pool1 = SqueezeNet.pool(conv1);
        const fire1 = SqueezeNet.create_fire_module(pool1, 16, 64);
        const fire2 = SqueezeNet.create_fire_module(fire1, 16, 64);
        const fire3 = SqueezeNet.create_fire_module(fire2, 32, 128);
        const pool2 = SqueezeNet.pool(fire3);
        const fire4 = SqueezeNet.create_fire_module(pool2, 32, 128);
        const fire5 = SqueezeNet.create_fire_module(fire4, 48, 192);
        const fire6 = SqueezeNet.create_fire_module(fire5, 48, 192);
        const fire7 = SqueezeNet.create_fire_module(fire6, 64, 256);
        const pool3 = SqueezeNet.pool(fire7);
        const fire8 = SqueezeNet.create_fire_module(pool3, 64, 256);
        const conv2 = SqueezeNet.create_conv(10, 1, fire8);
        const output = tf.layers.globalAveragePooling2d({});
        const model = tf.model(({inputs: input, outputs: output.apply(conv2)}));

         model.compile({
            optimizer: optimizer,
            loss: 'categoricalCrossentropy',
            metrics: ['accuracy'],
        });

        return model
    }

    static create_fire_module(model, squeeze_output, expand_output){
        const squeeze = SqueezeNet.create_conv(squeeze_output, 1, model);
        return SqueezeNet.create_expand(squeeze, expand_output)
    }

    static create_conv(output, kernel_size = 1, input = null){
        const conv = tf.layers.conv2d({
            kernelSize: [kernel_size, kernel_size],
            padding: 'same',
            filters: output,
            strides: 1,
            activation: 'relu',
            kernelInitializer: 'randomNormal',
            biasInitializer: 'zeros'
        });

        return (input) ? conv.apply(input) : conv
    }

    static create_expand(model, expand_output){
        const left = SqueezeNet.create_conv(expand_output, 1, model);
        const right = SqueezeNet.create_conv(expand_output, 3, model);
        const concatLayer = tf.layers.concatenate();
        return concatLayer.apply([left, right]);
    }

    static pool(model){
        const pool = tf.layers.maxPooling2d({poolSize: [2, 2], strides: [2, 2]});
        return pool.apply(model);
    }

    // Train the model.
    async train(set, set_size = BATCH_SIZE, train_steps = TRAIN_STEPS) {
        await this.model.fit(set.images.reshape([-1, IMAGE_SIZE, IMAGE_SIZE, IMAGE_DEPTH]),
                set.labels, {batchSize: BATCH_SIZE, epochs: train_steps});
        await tf.nextFrame();

    }

    // Predict the digit number from a batch of input images.
    async predict(batch, length){
        await this.model.predict(batch.images.reshape([-1 , IMAGE_SIZE, IMAGE_SIZE, IMAGE_DEPTH]),
            {batchSize: length});
    }

    /**
     * Shuffles training data and return subset of batch size
     * @return {{images[]; labels[]}}
     */
    nextBatch(type, size = BATCH_SIZE){

        console.log(`Info: Getting ${type} of size ${size}`);

        // let mapped = data[type].images.map((img, index) => {
        //         return {img: img, label: data[type].labels[index]}
        //     });
        //
        // console.log("Info: ", mapped);
        //
        // console.log(`Info: mapped data`);

        // const shuffled = mapped.sort(() => .5 - Math.random());// shuffle

        // console.log("Info: ", shuffled.length);
        //
        // console.log(`Info: shuffled data`);



        return {images: tf.tensor(data[type].images.slice(0, size-1)),
            labels: tf.tensor(data[type].labels.slice(0, size-1))}
    }
}

/*****************************
 *  SETUP
 ****************************/
export async function init(backend, training_size = TRAINING_SIZE, test_size = TEST_SIZE) {
    // Set backend to run on either CPU or GPU
    if(backend === 'gpu' || backend === 'cpu'){
        (backend === 'gpu') ? tf.setBackend('webgl') : tf.setBackend('cpu');
    } else {
        throw new Error(`Invalid backend parameter: ${backend}. Please specify either 'cpu' or 'gpu'`)
    }

    await loadData(training_size, test_size);
    return new SqueezeNet()
}

